File: form-runner/jvm/src/main/assets/xbl/orbeon/code-mirror/code-mirror.js
/**
 * Copyright (C) 2016 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
(function () {

  var $ = ORBEON.jQuery;

  ORBEON.xforms.XBL.declareCompanion('fr|code-mirror', {

    editor: null,
    handlers: {},

    hasFocus: false, // Heuristic: if has focus, don't update with value from server
    userChangedSinceLastBlur: false, // Use CodeMirror's own change event to track whether the value changed

    init: function () {

      var outer = $(this.container).find('.xbl-fr-code-mirror-editor-outer')[0];
      $('<xh:span class="xbl-fr-code-mirror-editor-inner"/>').appendTo(outer);

      var inner = $(this.container).find('.xbl-fr-code-mirror-editor-inner')[0];

      this.hasFocus = false;
      this.userChangedSinceLastBlur = false;
      
      this.editor = CodeMirror(
        inner,
        {
          mode: 'xml',
          lineNumbers: true,
          lineWrapping: true,
          indentUnit: 4,
          foldGutter: true,
          extraKeys: {"Ctrl-Q": function(cm){ cm.foldCode(cm.getCursor()); }},
          gutters: ["CodeMirror-linenumbers", "CodeMirror-foldgutter"]
        }
      );

      this.handlers = {
        'change': _.bind(this.codeMirrorChange, this), // TODO: .bind() is deprecated
        'focus': _.bind(this.codeMirrorFocus, this),
        'blur': _.bind(this.codeMirrorBlur, this)
      };

      var editor = this.editor;
      _.each(this.handlers, function (value, key) {
        editor.on(key, value);
      });

      this.xformsUpdateReadonly($(this.container).is('.xforms-readonly'));
    },
    destroy: function () {
      // Try to clean-up as much as we can
      var editor = this.editor;
      _.each(this.handlers, function (value, key) {
        editor.off(key, value);
      });
      this.handlers = {};
      this.editor = null;
      $(this.container).find('.xbl-fr-code-mirror-editor-outer').empty();
    },
    xformsFocus: function () {
      this.editor.focus();
    },
    codeMirrorFocus: function () {
      this.hasFocus = true;
    },
    codeMirrorBlur: function () {
      this.hasFocus = false;
      if (this.userChangedSinceLastBlur) {
        $(this.container).addClass('xforms-visited');
        ORBEON.xforms.Document.setValue(
          this.container.id,
          this.xformsGetValue()
          );
        this.userChangedSinceLastBlur = false;
      }
    },
    codeMirrorChange: function (codemirror, event) {
      if (event.origin != 'setValue') {
        this.userChangedSinceLastBlur = true;
      }
    },
    xformsUpdateReadonly: function (readonly) {
      if (readonly) {
        // Use 'true' instead of 'nocursor' so that copy/paste works:
        // https://github.com/orbeon/orbeon-forms/issues/1841
        this.editor.setOption('readOnly', 'true');
      } else {
        this.editor.setOption('readOnly', false);
      }
    },
    xformsUpdateValue: function (newValue) {
      var doUpdate =
        // As a shortcut, don't update the control if the user is typing in it
        !this.hasFocus &&
        // Don't update if the new value is the same as the current one, as doing so resets the editor position
        newValue != this.editor.getValue();
      if (doUpdate) {
        // It seems that we need a delay refresh otherwise sometimes the content doesn't appear when in a dialog
        // which is just shown. However, this doesn't work 100% of the time it seems.
        var editor = this.editor;

        var deferred = $.Deferred();

        setTimeout(function () {
          editor.setValue(newValue);
          deferred.resolve();
        }, 0);

        return deferred.promise();
      }
    },
    xformsGetValue: function () {
      return this.editor.getValue();
    }
  });
})();

File: form-runner-web/src/main/scala/org/orbeon/xbl/CodeMirror.scala
package org.orbeon.xbl

import org.orbeon.facades
import org.orbeon.xforms.facade.{XBL, XBLCompanion}
import org.orbeon.xforms.{$, AjaxClient, AjaxEvent, DocumentAPI}
import org.scalajs.dom.html
import org.scalajs.jquery.JQueryPromise

import scala.concurrent.{Future, Promise}
import scala.scalajs.js
import scala.scalajs.js.JSConverters._
import scala.scalajs.js.|
import org.scalajs.macrotaskexecutor.MacrotaskExecutor.Implicits._

object CodeMirror {

  XBL.declareCompanion("fr|code-mirror", js.constructorOf[CodeMirrorCompanion])

  private class CodeMirrorCompanion(containerElem: html.Element) extends XBLCompanion {

    private var editor                   : facades.CodeMirror = null
    private var handlers                 : js.Dictionary[js.Function] = js.Dictionary()
    private var hasFocus                 : Boolean = false
    private var userChangedSinceLastBlur : Boolean = false

    override def init(): Unit = {
      val editorElement = containerElem.querySelector(".xbl-fr-code-mirror-editor")
      this.editor = new facades.CodeMirror(
        editorElement,
        js.Dictionary(
          "mode"         -> "xml",
          "lineNumbers"  -> true,
          "lineWrapping" -> true,
          "indentUnit"   -> 4,
          "theme"        -> "darcula",
          "foldGutter"   -> true,
          "gutters"      -> js.Array("CodeMirror-linenumbers", "CodeMirror-foldgutter"),
          "extraKeys"    -> js.Dictionary(
            "Ctrl-Q"     -> ((cm: js.Dynamic) => cm.foldCode(cm.getCursor())),
            "Cmd-Enter"  -> js.Any.fromFunction0(userCmdCtrlEnter _),
            "Ctrl-Enter" -> js.Any.fromFunction0(userCmdCtrlEnter _)
          )
        )
      )

      this.handlers = js.Dictionary(
        "change" -> this.codeMirrorChange _,
        "focus"  -> this.codeMirrorFocus _,
        "blur"   -> this.codeMirrorBlur _
      )

      this.handlers.foreach { case (key, value) => this.editor.on(key, value) }

      this.xformsUpdateReadonly($(containerElem).is(".xforms-readonly"))
    }

    override def destroy(): Unit = {
      this.handlers.foreach { case (key, value) => this.editor.off(key, value) }
      this.handlers.clear()
      this.editor = null
      $(containerElem).find(".xbl-fr-code-mirror-editor").empty()
    }

    override def xformsFocus(): Unit = this.editor.focus()

    private def codeMirrorFocus(): Unit = this.hasFocus = true

    private def userCmdCtrlEnter(): Unit = {
      sendValueToXForms() foreach { _ =>
        AjaxClient.fireEvent(
          AjaxEvent(
            eventName = "DOMActivate",
            targetId  = this.containerElem.id
          )
        )
      }
    }

    // We update the value on blur, not on change, to be incremental, for performance on large forms
    private def codeMirrorBlur(): Unit = {
      this.hasFocus = false
      if (this.userChangedSinceLastBlur) {
        sendValueToXForms()
        this.userChangedSinceLastBlur = false
      }
    }

    private def sendValueToXForms(): Future[Unit] = {
      $(containerElem).addClass("xforms-visited")
      DocumentAPI.setValue(
        containerElem.id,
        this.xformsGetValue()
      )
    }

    private def codeMirrorChange(codeMirror: js.Dynamic, event: js.Dynamic): Unit = {
      val eventOrigin = event.origin.asInstanceOf[js.UndefOr[String]]
      if (! eventOrigin.contains("setValue")) {
        this.userChangedSinceLastBlur = true
      }
    }

    override def xformsUpdateReadonly(readonly: Boolean): Unit = {
      if (readonly) this.editor.setOption("readOnly", "true")
      else          this.editor.setOption("readOnly", false)
    }

    override def xformsUpdateValue(newValue: String): js.UndefOr[js.Promise[Unit] | JQueryPromise] = {
      val doUpdate = ! this.hasFocus && newValue != this.editor.getValue()
      if (doUpdate) {
        val editor = this.editor
        val promise = Promise[Unit]()

        js.timers.setTimeout(0) {
          editor.setValue(newValue)
          promise.success(())
        }

        promise.future.toJSPromise
      } else {
        js.undefined
      }
    }

    override def xformsGetValue(): String = this.editor.getValue()
  }
}


